cpp/src/c_api/graph_mg.cpp

```diff
@@ -298,87 +297,198 @@ extern "C" cugraph_error_code_t cugraph_mg_graph_create(
   auto p_dst =
-    reinterpret_cast<cugraph::c_api::cugraph_type_erased_device_array_view_t const*>(dst);
+    reinterpret_cast<cugraph::c_api::cugraph_type_erased_device_array_view_t const* const*>(dst);
   auto p_weights =
-    reinterpret_cast<cugraph::c_api::cugraph_type_erased_device_array_view_t const*>(weights);
+    reinterpret_cast<cugraph::c_api::cugraph_type_erased_device_array_view_t const* const*>(
+      weights);
   auto p_edge_ids =
-    reinterpret_cast<cugraph::c_api::cugraph_type_erased_device_array_view_t const*>(edge_ids);
+    reinterpret_cast<cugraph::c_api::cugraph_type_erased_device_array_view_t const* const*>(
+      edge_ids);
   auto p_edge_type_ids =
-    reinterpret_cast<cugraph::c_api::cugraph_type_erased_device_array_view_t const*>(edge_type_ids);
+    reinterpret_cast<cugraph::c_api::cugraph_type_erased_device_array_view_t const* const*>(
+      edge_type_ids);
+
+  size_t local_num_edges{0};
+
+  //
+  // Determine the type of vertex, weight, edge_type_id across
+  // multiple input arrays and acros multiple GPUs.  Also compute
+  // the number of edges so we can determine what type to use for
+  // edge_t
+  //
+  cugraph_data_type_id_t vertex_type{cugraph_data_type_id_t::NTYPES};
+  cugraph_data_type_id_t weight_type{cugraph_data_type_id_t::NTYPES};
+
+  for (size_t i = 0; i < num_arrays; ++i) {
+    CAPI_EXPECTS(p_src[i]->size_ == p_dst[i]->size_,
+                 CUGRAPH_INVALID_INPUT,
+                 "Invalid input arguments: src size != dst size.",
+                 *error);
+
+    CAPI_EXPECTS(p_src[i]->type_ == p_dst[i]->type_,
+                 CUGRAPH_INVALID_INPUT,
+                 "Invalid input arguments: src type != dst type.",
+                 *error);
+
+    CAPI_EXPECTS((p_vertices == nullptr) || (p_src[i]->type_ == p_vertices[i]->type_),
+                 CUGRAPH_INVALID_INPUT,
+                 "Invalid input arguments: src type != vertices type.",
+                 *error);
+
+    CAPI_EXPECTS((weights == nullptr) || (p_weights[i]->size_ == p_src[i]->size_),
+                 CUGRAPH_INVALID_INPUT,
+                 "Invalid input arguments: src size != weights size.",
+                 *error);
+
+    local_num_edges += p_src[i]->size_;
+
+    if (vertex_type == cugraph_data_type_id_t::NTYPES) vertex_type = p_src[i]->type_;
+
+    if (weights != nullptr) {
+      if (weight_type == cugraph_data_type_id_t::NTYPES) weight_type = p_weights[i]->type_;
+    }
 
-  CAPI_EXPECTS(p_src->size_ == p_dst->size_,
-               CUGRAPH_INVALID_INPUT,
-               "Invalid input arguments: src size != dst size.",
-               *error);
-  CAPI_EXPECTS(p_src->type_ == p_dst->type_,
+    CAPI_EXPECTS(p_src[i]->type_ == vertex_type,
+                 CUGRAPH_INVALID_INPUT,
+                 "Invalid input arguments: all vertex types must match",
+                 *error);
+
+    CAPI_EXPECTS(p_weights[i]->type_ == weight_type,
```

Comment:
You forgot to check if `weights` was not a `nullptr` otherwise we get a `segfault`. I changed it locally to this

```
CAPI_EXPECTS((weights == nullptr) || (p_weights[i]->type_ == weight_type),
                 CUGRAPH_INVALID_INPUT,
                 "Invalid input arguments: all weight types must match",
                 *error);
```